components/feature/pwa/src/main/java/mozilla/components/feature/pwa/feature/ManifestUpdateFeature.kt

```diff
@@ -0,0 +1,72 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.pwa.feature
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.concept.engine.manifest.WebAppManifest
+import mozilla.components.feature.pwa.ManifestStorage
+import mozilla.components.feature.pwa.WebAppShortcutManager
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+
+/**
+ * Feature used to update the existing web app manifest and web app shortcut.
+ *
+ * @param shortcutManager Shortcut manager used to update pinned shortcuts.
+ * @param storage Manifest storage used to have updated manifests.
+ * @param sessionId ID of the web app session to observe.
+ * @param initialManifest Loaded manifest for the current web app.
+ */
+class ManifestUpdateFeature(
+    private val applicationContext: Context,
+    private val sessionManager: SessionManager,
+    private val shortcutManager: WebAppShortcutManager,
+    private val storage: ManifestStorage,
+    private val sessionId: String,
+    private var initialManifest: WebAppManifest
+) : Session.Observer, LifecycleAwareFeature {
+
+    private var scope: CoroutineScope? = null
+    private var updateJob: Job? = null
+
+    /**
+     * When the manifest is changed, compare it to the existing manifest.
+     * If it is different, update the disk and shortcut. Ignore if called with a null
+     * manifest or a manifest with a different start URL.
+     */
+    override fun onWebAppManifestChanged(session: Session, manifest: WebAppManifest?) {
+        if (manifest?.startUrl == initialManifest.startUrl && manifest != initialManifest) {
```

Comment:
How do we want to deal with the case where the start URL changed in the future? Is that somewhat defined?

Reply:
Right now we don't have a way to update the shortcuts if the start URL changes, since the ID of the shortcut will be different. The fix is to re-install the PWA.